content/posts/creating-grafana-dashboards-for-nodejs-on-kubernetes/index.md

```diff
@@ -0,0 +1,261 @@
+The following query calculates the rate for each bucket and pod and then sums that up by bucket to produce the chart above.
+
+```promql
+sum by (le) (rate(http_request_duration_seconds_bucket{kubernetes_namespace="awesome-node"}[3m]))
+```
+
+A drawback of this chart is, that most of the requests are in the bucket `le: 0.003` and the graphs representing other buckets all get crammed into a very tiny space.
+
+We can exclude the bucket for the shortest requests and remove them from the sum by turning the buckets around and calculating the request counts that are larger than the `le` thresholds.
+
+![Graph chart showing request buckets excluding the fastest bucket](./request-duration-stacked_reduced.png)
+
+Use the query below to produce this chart:
+
+```promql
+sum by (app, le) (rate(http_request_duration_seconds_count{app="awesome-node"}[3m]) - ignoring (le) group_right rate(http_request_duration_seconds_bucket{le!="+Inf"}[3m]))
+```
+
+> #### Query Explanation
+>
+> This query uses [one-to-many](https://prometheus.io/docs/prometheus/latest/querying/operators/#many-to-one-and-one-to-many-vector-matches) matching to subtract many metrics (the buckets) from a single metric (the count). In order to support this, we need to tell Prometheus that the right part has the higher cardinality using the `group_right` modifier.
+>
+> Additionally, Prometheus by default matches metrics whose labels also match. However, the `le` label is only available on the buckets. For the matching operation to succeed, we tell it to ignore the `le` label, by adding the `ignoring (le)` modifier.
+
+### Histogram Heatmap
+
+Another interesting way to visualize histograms are heatmaps and Grafana's heatmap implementation is well suited for it.
+
+![Heatmap showing request rate per bucket over 40 minutes](./heatmap.png)
+
+As you can see on this heatmap, the amount of long requests is clearly shown in the upper parts of the heatmap with a more intense colour indicating more requests in that bucket.
+Optimally, only the bottom 1-3 rows should be colored at all, with few exceptions.
+
+The query to produce this heatmap is quite simple:
+
+```promql
+label_replace(sum by (le) (increase(http_request_duration_seconds_bucket{app="awesome-node"}[$__rate_interval])), "le", "Inf", "le", "\\+Inf")
+```
+
+> Note!
+>
+> Using absolute numbers instead of rates per second can be practical when grouping into larger intervals such as in this use case.
+>
+> The `label_replace()` function is used here to replace the `le: +Inf` label with `le: Inf`, because Grafana Would use the `+Inf` label incorrectly.
+>
+> We're using the $\_\_rate_interval instead of a fixed timespan such as `3m` to ensure all data points are counted only once and within the correct heatmap section.
+
+The metric needs to be configured as a heatmap in Grafana and in this case it is useful to scale down the resolution of the heatmap.
+
+![Metric configuration for the heatmap](./heatmap-config-metric.png)
+
+> Important!
```

Comment:
I would just use `#### Important` here to make it more syntactically correct.

Reply:
Do you mean without the blockquote? Would you recommend the same for the note blockquote above?